Issue #5129 - WebAppContext.setExtraClasspath(String) cleanup

joakime opened the pull request:
+ More tests for both relative and absolute path references
+ More testing that will trigger quirks on Windows builds
  so that we can catch regressions faster
+ Reworked WebInfConfiguration to be glob aware in a way
  similar to how WebAppClassLoader behaves.
+ Reworked Resource.newResource(String) to delegate
  canonical path resolution to PathResource
+ Guarded PathResource's usage of Path.toAbsolutePath()
  to ignore valid conditions where the Path cannot be
  resolved to an absolute path (yet)

joakime committed:
Issue #5129 - WebAppContext.setExtraClasspath(String) cleanup

+ More tests for both relative and absolute path references
+ More testing that will trigger quirks on Windows builds
  so that we can catch regressions faster
+ Reworked WebInfConfiguration to be glob aware in a way
  similar to how WebAppClassLoader behaves.
+ Reworked Resource.newResource(String) to delegate
  canonical path resolution to PathResource
+ Guarded PathResource's usage of Path.toAbsolutePath()
  to ignore valid conditions where the Path cannot be
  resolved to an absolute path (yet)
+ Normalize resolved paths in PathResource

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

janbartel commented on jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java:
```diff
@@ -947,13 +947,47 @@ protected static String getResourceBaseName(Resource resource)
-            int dot = fnlc.lastIndexOf('.');
-            String extension = (dot < 0 ? null : fnlc.substring(dot));
-            if (extension != null && (extension.equals(".jar") || extension.equals(".zip")))
+            String token = tokenizer.nextToken().trim();
```
I think the problem is that we are parsing this extraClassPath String in 2 places: once in WebAppLoader, but then again in WebInfConfiguration. It should only be parsed once. So I think the better solution - as it is WebInfConfiguration's job to set up the WebAppLoader - that it converts the extraClassPath into Resources, and then sets those resources onto WebAppLoader  in the configure() method, where it pushes the WEB-INF/lib jars and WEB-INF/classes resources into the WebAppLoader.

joakime replied:
There's actually 3 places.

gregw replied:
Will be cleaned up in 10

janbartel commented on jetty-webapp/src/test/java/org/eclipse/jetty/webapp/WebAppContextExtraClasspathTest.java:
```diff
@@ -0,0 +1,216 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class WebAppContextExtraClasspathTest
+{
```
I think all these tests are good, but can they be added to an existing test, like for WebAppClassLoader, or WebAppContext?

joakime replied:
I wanted to test extraClassPath in isolation.
We have 2 separate WebAppClassLoader test classes. (which has good separation)
And we have a giant multi-purpose WebAppContext test that does far too many things all at once.  It should really be broken down into more isolated tests.

joakime committed:
Issue #5129 - Simplify Resource reference list behavior

+ Introduce new Resource.fromReferences to help with
  parsing delimited resource reference lists.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5129 - Resource.fromReferences no longer needs DefaultFactory

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5129 - Fleshing out glob testcase a bit more

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw reviewed (changes requested):
This PR looks a bit fundamental for jetty-9.4!  Is there a simpler fix we can do for the bug in 9.4 and then consider code cleanup in 10.  I've not yet looked at this in detail but the the `ResourceFactory` vs `Resource.Factory` stuff seams really strange.

So let's just fix the bug here and do a cleanup/restructure in 10.

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:
```diff
@@ -174,19 +177,8 @@ public static Resource newResource(String resource, boolean useCaches)
-                    throw e2;
-                }
+                // It's likely a file/path reference.
+                return new PathResource(Paths.get(resource));
```
This might be a better assumption, but for 9.4 I think we really don't want to change how we work out what type a resource is as we really don't want to change behaviour... even if it is slightly wrong.  We should only fix real bugs.

joakime replied:
The code that was removed was a left-over from the FileResource days.

The use of getCanonicalFile() is one of the core issues from #5129 and must be addressed.
On windows, it throws for many reasons, during testing I found that it throws if the File (or any sub-path it references) doesn't exist or it contains characters in the list of unacceptable characters (such as the `*` in `path\to\dir\*`).  Note: these 2 reasons are not the only reasons getCanonicalFile throws.   (On a typical Linux environment I have been unable to produce a combination of conditions that would cause it to throw)

The stripping of the "./" prefix is something that the `Path.normalize()` takes care of.
The modified form of the `resource` variable that the old lines 180-181 produces never exits this block of code, you'll have a `PathResource` or an exception.

janbartel reviewed (changes requested):
This is a very big change for 9.4, so I think the bulk of this improvement work should be done in jetty-10. For 9.4, we can just fix the 2 problems:  
1.  WebInfConfiguration needs to strip off a trailing * in parsing extraClassPath
2. Resource needs to handle weird NTFS better

Then, in jetty-10, lets consider how to prevent the triplicate parsing of the extraClassPath. The solution with having the Resource.Factory interface is moving in the right direction, but I really don't like having that AND Resource.Factory : the benefit of doing this in 10 is that we will be able to change the signature of ResourceFactory.  Also, this solution doesn't overcome the triplicate parsing, it's still being parsed twice in WebInfConfiguration and once in WebAppClassLoader : lets just parse this once and use the results where needed.

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:
```diff
@@ -1005,4 +996,107 @@ public static URL toURL(File file) throws MalformedURLException
+     * @param resourceFactory the Resource.Factory used to create new Resource references
+     * @return the list of resources parsed from input string.
+     */
+    public static List<Resource> fromReferences(String delimitedReferences, Resource.Factory resourceFactory) throws IOException
```
`fromList` or `fromCSV` would be a better name, as `References` doesn't mean anything to me.

joakime replied:
`fromList` I'm good with, `fromCSV` I'm against, as this doesn't handle CSV, merely delimited. (I would also accept `fromDelimited`)
None of our existing methods with CSV in the name actually handle CSV. (and that bothers me)

joakime replied:
This layer has been removed in this PR for `jetty-9.4.x`.
It should show up again in a new PR for `jetty-10.0.x`

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:
```diff
@@ -1005,4 +996,107 @@ public static URL toURL(File file) throws MalformedURLException
+         * @return the Resource instance
+         * @throws IOException if unable to create a Resource reference
+         */
+        Resource newResource(String reference) throws IOException;
```
Do we really need a new public interface just because `ResourceFactory` can return null?    It's only called in a few spots in the methods below and null checking there would be better than introducing a nearly identical interface..... or just use `Function<String, Resource>`

We could perhaps change `ResourceFactory` in 10 to throw, but that is a lot more expensive for 404s

joakime replied:
The pre-existing `ResourceFactory` is really a `ResourceAddPathFactory` or `SubResourceFactory`, it's badly named.

I would love to use something like `Function<String,Resource>` but that doesn't allow for throwing of the exception.

joakime replied:
If we can skip the `Context.newResource(String)` layer for extraClassPath parsing then the `Resource.Factory` can be removed.

joakime replied:
This layer has been removed in this PR for `jetty-9.4.x`.
It should show up again in a new PR for `jetty-10.0.x`

gregw reviewed (changes requested):
On looking closer, I'm not so sure this needs to be entirely punted to 10.    I think it can be greatly simplified without the introduction of new public interfaces.      If that can be done, then perhaps this can go in 9.4, but any significant API changes can only be done in 10... and then only really soon.

joakime commented:
> This is a very big change for 9.4, so I think the bulk of this improvement work should be done in jetty-10. For 9.4, we can just fix the 2 problems:
> 
> 1. WebInfConfiguration needs to strip off a trailing * in parsing extraClassPath
> 2. Resource needs to handle weird NTFS better

This cannot be done entirely within the Resource layer, it has to be done BEFORE the entry becomes a Resource.

> Then, in jetty-10, lets consider how to prevent the triplicate parsing of the extraClassPath. The solution with having the Resource.Factory interface is moving in the right direction, but I really don't like having that AND Resource.Factory : the benefit of doing this in 10 is that we will be able to change the signature of ResourceFactory. Also, this solution doesn't overcome the triplicate parsing, it's still being parsed twice in WebInfConfiguration and once in WebAppClassLoader : lets just parse this once and use the results where needed.

I see 2 paths we can take ...

_Option 1:_

The best we can do is reduce it to 2 parsings.

Once in the optional WebInfConfiguration, and once in the WebAppClassLoader.addClassPath(String) method.

As people use WebAppContext with a dizzying variety of Configurations (and often no Configurations in various embedded scenarios, when they could easily use ServletContextHandler instead).
And people use WebAppClassLoader by directly manipulating it.

_Option 2:_

We don't store `WebAppContext.setExtraClassPath(String)` as a String.

The immediate usage of that method results in a `List<Resource>` and is resolved immediately.

We  introduce a new `WebAppContext.setExtraClassPath(List<Resource>)` method, and update the apidoc for above method to say it's just a convenience method for this one.
The `.getExtraClassPath()` method return is changed to `List<Resource>` instead.

This has a side effect that the `Context.newResource(String)` methods are not used though.

joakime committed:
Issue #5129 - Reverting Resource.fromReferences()

+ Will migrate to jetty-10.0.x

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #5129 - disabling ResourceTest.testGlobPath on WINDOWS for stability

janbartel commented on jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java:
```diff
@@ -319,29 +319,37 @@ public void addJars(Resource lib)
+                            LOG.debug("addJar - {}", resource);
+                        if (resource.isDirectory())
+                        {
+                            addURL(resource.getURI().toURL());
```
Wwe've ever supported an extraClasspath directory hierarchy that is a mix of classes dirs and jar files. The arg to addJars() has always been interpreted as a dir that contains jar/zip files, and only the jar/zip files were added to the classpath. It looks like line 334 is going to add the subdirs of the lib to the classpath, but that's never been the accepted behaviour.

joakime replied:
No, it doesn't add sub-directories, it only adds the directory itself as a directory reference.
Similar to how WebInfConfiguration.findExtraClassPathDirs() works with glob.

janbartel commented on jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java:
```diff
@@ -1003,11 +1037,53 @@ protected Resource findWebInfClassesDir(WebAppContext context)
-                dirResources.add(resource);
+            String token = tokenizer.nextToken().trim();
+            if (isGlobReference(token))
+            {
```
If the entry has a trailing *, then it is a directory of jar files and we can ignore it, we should not look deeply into its subdirs.

joakime replied:
We only look at the level of the `/*`.
So that means if we have say ...

``` shell
$ ls -1 -F /opt/ext/
corp.jar
foo.jar
resources/
```

And a reference of `/opt/ext/*` then that's 3 URL entries in the WebAppClassLoader.

1. file:///opt/ext/corp.jar  
2. file:///opt/ext/foo.jar
3. file:///opt/ext/resources/  (treated as directory entry in the ClassLoader)

janbartel replied:
This is not the WebAppClassLoader, this is in WebInfConfiguration.findExtraClasspathDirs where we are only looking for entries in the ;, separated list that are themselves directories; any entry that is suffixed by /* is handled by WebInfConfiguration.findExtraClasspathJars. We don't want to do a deep add of a directory inside of WebInfConfiguration.findExtraClasspathDirs.

joakime replied:
This doesn't do deep directory, it's only at the level of the glob.

Glob means anything relevant in the point of the tree of the glob, not deeper.

"Relevant" is anything that has meaning to the Classloader, which is zip, jar, and directory (all other entries that the glob produces are ignored).

joakime replied:
So, we have a misaligned behavior here.

`WebInfConfiguration` and `WebAppClassLoader` in existing jetty-9.4.x behaves differently in regards to `/path/to/*` entries.

We either use the `WebAppClassLoader` behavior for glob that adds directories.
Or we use the `WebInfConfiguration` behavior for glob that **does not** add directories.

Looking at how java itself works I think we should use the existing `jetty-9.4.x` behavior that `WebInfConfiguration` uses and not add directories.

I'll explain with a code example ...

``` java
package cloader;

import java.io.File;

public class ShowClassLoader
{
    public static void main(String[] args)
    {
        String javaClasspath = System.getProperty("java.class.path");
        System.out.printf("java.class.path=%s%n", javaClasspath);
        String[] parts = javaClasspath.split(File.pathSeparator);
        System.out.printf("  entries.length=%d%n", parts.length);
        for (int i = 0; i < parts.length; i++)
        {
            System.out.printf("  entry[%d]=%s%n", i, parts[i]);
        }
    }
}
```

Run on the command line ...

``` shell
$ ls -1 -F target/lib
corp.jar
foo.jar
resources/

$ java -cp 'target/lib/*:target/classes' cloader.ShowClassLoader
java.class.path=target/lib/corp.jar:target/lib/foo.jar:target/classes
  entries.length=3
  entry[0]=target/lib/corp.jar
  entry[1]=target/lib/foo.jar
  entry[2]=target/classes
```

This behavior on OpenJDK 11.0.7 is what `WebInfConfiguration` mimics, but not what `WebAppClassLoader` does.

The OpenJDK 11.0.7 on Windows classpath behaves the same way.

And it's documented like this at https://docs.oracle.com/javase/8/docs/technotes/tools/windows/classpath.html#A1100762

joakime replied:
Just as a followup, the glob behaviors on Linux from the above example seem to be mostly done via the shell.

Windows JVM seems to have specific code to allow globs, so we are essentially attempting to mimic glob shell behavior on linux and glob classpath behavior on windows.

joakime replied:
This has been reverted in WebInfConfiguration and WebAppClassloader has been updated to reflect the truth in both JVM behavior and WebInfConfiguration behavior (along with tests).

janbartel replied:
So the point I was making was that in WebInfConfiguration.findExtraClasspathJars we should not add the directory with the trailing * to the list of resources, only the jars and zips that are contained at the top level inside of that directory. If you've changed the code so that that directory is _not_ added, then we're good.

joakime replied:
Current implementation in this PR does not add directories for glob usage in either WebInfConfiguration (like before) nor in WebAppClassLoader (this is a change in behavior)

janbartel commented on jetty-webapp/src/test/java/org/eclipse/jetty/webapp/WebAppContextExtraClasspathTest.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    @Test
+    public void testBaseResourceAbsolutePath() throws Exception
```
This test needs a better place, it is not testing the extraClasspath.

joakime replied:
That would mean a new class.

As the existing WebAppContextTest forces behavior on each run that is irrelevant for this test.
I can clean that up in the future `jetty-10.0.x` PR though.

joakime replied:
I opened Issue #5137 to cleanup the WebAppContext tests.
Lets keep this PR on the smallish side.

gregw replied:
One of my pet peeves is that we create too many special purpose test classes and it can be really hard to find where things are tested or what tests to run to get full coverage for a change.  As much as possible we should keep unit tests for Foo.java in FooTest.java

joakime committed:
Issue #5129 - Directories discovered via classpath glob are not valid

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

janbartel reviewed (approved):
lgtm

KevinMcT commented:
Hi, I just wanted to note that this PR (Sparked from the issue https://github.com/eclipse/jetty.project/issues/5129) ended up as a breaking change in our system which has been running stable without issue for the past 7 years.

(For future reference.)

```
java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/Users/xxx/.gradle/caches/modules-2/files-2.1/org.constretto/constretto-api/2.2.3/7e7a08dcb9abdfc4fb2a610a6edcad1bb40c1444/constretto-api-2.2.3.jar
	at sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at sun.nio.fs.WindowsPath.parse(WindowsPath.java:94)
	at sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:255)
	at java.nio.file.Paths.get(Paths.java:84)
	at org.eclipse.jetty.util.resource.Resource.newResource(Resource.java:179)
	at org.eclipse.jetty.util.resource.Resource.newResource(Resource.java:152)
	at org.eclipse.jetty.server.handler.ContextHandler.newResource(ContextHandler.java:2027)
	at org.eclipse.jetty.webapp.WebAppClassLoader.addClassPath(WebAppClassLoader.java:272)
	at org.eclipse.jetty.webapp.WebAppClassLoader.<init>(WebAppClassLoader.java:195)
	at org.eclipse.jetty.webapp.WebAppClassLoader.<init>(WebAppClassLoader.java:158)
	at org.eclipse.jetty.webapp.WebAppContext.preConfigure(WebAppContext.java:467)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:523)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.server.handler.StatisticsHandler.doStart(StatisticsHandler.java:253)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
	at org.eclipse.jetty.server.Server.start(Server.java:423)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
	at org.eclipse.jetty.server.Server.doStart(Server.java:387)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:72)
```

While testing an upgrade from jetty 9.4.31 -> 9.4.32 it was discovered that this issue propped up for dev's using Windows. Issue is not present for dev's with Linux/MacOS machines.

URI as given in the URI object:

```
file:/C:/Users/xxx/.gradle/caches/modules-2/files-2.1/org.constretto/constretto-api/2.2.3/7e7a08dcb9abdfc4fb2a610a6edcad1bb40c1444/constretto-api-2.2.3.jar
```

```
            // -> Yes: Give the classpath to Jetty
            String classpath = ClasspathHelper.convertSetToClasspathString(_webAppClassPath);
            log.info("Webapp classpath: " + classpath);
            context.setExtraClasspath(classpath);
```
The ClasspathHelper method call is nothing fancy, does this:
```
static String convertSetToClasspathString(Set<URI> classpath) {
        return classpath.stream().map(URI::getPath).collect(Collectors.joining(";"));
    }
```

The URI::getPath strips off the `file:` prefix, the returned string is then feed into the setExtraClasspath method.

We remedied the issue by rewriting the stream from utilizing getPath to toString
```
String classpath = _webAppClassPath.stream().map(URI::toString).collect(Collectors.joining(";"));
```

This wasn't a major issue, however introducing a breaking change in a patch release is neither expected or desired.

gregw commented:
@KevinMcT  Sorry about that, it was not our intention.  I have opened #5871 as that behaviour sounds like a bug to me.

gregw commented:
@KevinMcT after discussing this with the team, we are not so sure that this is something we can fix.  A string like `/C:/some/path` is neither a URI nor a file path and thus should not be supported by the setExtraClasspath method.    If anything, the fact that it may have previously been supported was a bug that needed to be fixed... which we may have done inadvertently with previous changes.
Thus I'm going to mark #5871 as closed for now.  If you can think of a reason we should support  `/C:/some/path`, then please reopen and explain.

KevinMcT commented:
@gregw np! I don't disagree with your conclusion. My main point was to highlight the fact that a change was introduced which might hit other users as well. (Leaving the note for future reference.)

Seeing that even though this was a bug which was fixed inadvertently, while reading through the release notes it wasn't immediately obvious that this was going to end up as a breaking change (for us that is).
